pkg/stanza/fileconsumer/internal/tracker/tracker.go

```diff
@@ -0,0 +1,200 @@
+
+type Tracker struct {
+	*zap.SugaredLogger
+
+	ReaderFactory reader.Factory
+	maxBatchFiles int
+
+	currentPollFiles  *fileset.Fileset[*reader.Reader]
+	previousPollFiles *fileset.Fileset[*reader.Reader]
+	knownFiles        []*fileset.Fileset[*reader.Metadata]
+}
+
+type trackerOp func(tracker *Tracker)
+
+func New(logger *zap.SugaredLogger, readerFactory reader.Factory, maxBatchFiles int) *Tracker {
+	knownFiles := make([]*fileset.Fileset[*reader.Metadata], 3)
+	for i := 0; i < len(knownFiles); i++ {
+		knownFiles[i] = fileset.New[*reader.Metadata](maxBatchFiles)
+	}
+	return &Tracker{
+		SugaredLogger:     logger.With("component", "fileconsumer"),
+		ReaderFactory:     readerFactory,
+		maxBatchFiles:     maxBatchFiles,
+		currentPollFiles:  fileset.New[*reader.Reader](maxBatchFiles),
+		previousPollFiles: fileset.New[*reader.Reader](maxBatchFiles),
+		knownFiles:        knownFiles,
+	}
+}
+
+func (t *Tracker) PostPoll(persister operator.Persister) {
+	// Any new files that appear should be consumed entirely
+	t.ReaderFactory.FromBeginning = true
+	t.Persist(persister)
+	// rotate at end of every poll()
+	t.rotateFilesets()
+}
+
+func (t *Tracker) OpenFile(path string) {
+	fmt.Println(t.currentPollFiles.Len(), t.previousPollFiles.Len())
+	// if t.currentPollFiles.Len()+t.previousPollFiles.Len() > t.maxBatchFiles*2 {
+	// 	t.Errorw("cannot open file", zap.Error(errTooManyActiveFiles))
+	// 	return
+	// }
+	fp, file := t.makeFingerprint(path)
+	if fp == nil {
+		return
+	}
+
+	// Exclude duplicate paths with the same content. This can happen when files are
+	// being rotated with copy/truncate strategy. (After copy, prior to truncate.)
+	if r := t.currentPollFiles.Match(fp, fileset.Equal); r != nil {
+		// re-add the reader as Match() removes duplicates
+		t.currentPollFiles.Add(r)
+		if err := file.Close(); err != nil {
+			t.Debugw("problem closing file", zap.Error(err))
+		}
+		return
+	}
+
+	r, err := t.newReader(file, fp)
+	if err != nil {
+		t.Errorw("Failed to create reader", zap.Error(err))
+		return
+	}
+
+	t.currentPollFiles.Add(r)
+}
+
+func (t *Tracker) CurrentPollFiles() *fileset.Fileset[*reader.Reader] {
+	return t.currentPollFiles
+}
+
+func (t *Tracker) PreviousPollFiles() *fileset.Fileset[*reader.Reader] {
+	return t.previousPollFiles
+}
+
+func (t *Tracker) Persist(persister operator.Persister) {
+	if persister != nil {
+		allCheckpoints := make([]*reader.Metadata, 0, t.TotalReaders())
+		for _, knownFiles := range t.knownFiles {
+			allCheckpoints = append(allCheckpoints, knownFiles.Get()...)
+		}
+
+		for _, r := range t.previousPollFiles.Get() {
+			allCheckpoints = append(allCheckpoints, r.Metadata)
+		}
+		if err := checkpoint.Save(context.Background(), persister, allCheckpoints); err != nil {
+			t.Errorw("save offsets", zap.Error(err))
+		}
+	}
+}
+
+func (t *Tracker) Load(ctx context.Context, persister operator.Persister) error {
```

Comment:
Similarly, rather than having the tracker load the files, we should keep this in the manager and pass the loaded files into the tracker during initialization.